### PR TITLE
reorganize the higher order autodiff api

### PR DESCRIPTION
**Pull request by cxxly**
### PR types

Others
### PR changes

APIs
### Describe

- add reverse mode autodiff api `paddle.incubate.autograd.grad`, which internally use `paddle.static.gradients` when disable_prim, and use linearize/transpose transforms when enable_prim.
- rename forward mode autodiff api  `paddle.incubate.autograd.forward_graidents` to `paddle.incubate.autograd.forward_grad`, and make it public.
- make `paddle.incubate.autograd.prim_enabled` and `paddle.incubate.autograd.prim2orig` unpublic.
- change `paddle.incubate.autograd.jvp` to call `paddle.incubate.autograd.forward_grad` when enable_prim.
- remove remaining jacobian/hessian/batch_jacobian/batch_hessian/hvp api.
- fix corresponding unittest code..

### CODE REVIEW

File: python/paddle/fluid/tests/unittests/autograd/test_autograd_functional_prim.py [hunk 1]
```diff
@@ -83,67 +83,66 @@ def wrapper(fun, args):
...
+# @utils.place(config.DEVICES)
+# @utils.parameterize((utils.TEST_CASE_NAME, 'fun', 'args', 'dtype'), (
+#     ('unary_float32', paddle.tanh, (np.random.rand(1), ), 'float32'),
+#     ('binary_float32', paddle.multiply,
+#      (np.random.rand(1), np.random.rand(1)), 'float32'),
+#     ('unary_float64', paddle.tanh, (np.random.rand(1), ), 'float64'),
+#     ('binary_float64', paddle.multiply,
+#      (np.random.rand(1), np.random.rand(1)), 'float64'),
+# ))
+# class TestHessianPrim(unittest.TestCase):
```
Comment (levi131) [inline python/paddle/fluid/tests/unittests/autograd/test_autograd_functional_prim.py]: 这里注释掉测例是出于什么考虑呢

File: python/paddle/incubate/autograd/functional.py [hunk 2]
```diff
@@ -0,0 +1,702 @@
...
+        return paddle.concat(tuple(x.reshape((-1, )) for x in xs))
+
+    def _evaluate(self, row_index):
+        return self._flatten(_grad(
+            self._flatten_ys[row_index],
+            self._xs,
+        ))
+
+
+class _JacobianBatchLast(_Jacobian):
```
Comment (levi131) [inline python/paddle/incubate/autograd/functional.py]: 这个batchlast 拿掉吧，我们只用到batchfirst

File: python/paddle/incubate/autograd/functional.py [hunk 3]
```diff
@@ -0,0 +1,702 @@
...
+    _check_inputs(func, xs, v)
+    # ``_seprate`` breaks the dependencies between ``xs`` and other
+    # variables. See more ``_seprate`` .
+    if paddle.fluid._non_static_mode() or not utils.prim_enabled():
+        xs, v = _separate(xs), _separate(v)
+    ys = func(*xs) if isinstance(xs, typing.Sequence) else func(xs)
+    _check_v_shape(v, xs)
+
+    if not paddle.fluid._non_static_mode() and utils.prim_enabled():
+        return ys, primx.forward_grad(ys, xs, v)
```
Comment (levi131) [inline python/paddle/incubate/autograd/functional.py]: primx.forward_grad是不是应该改为primapi.forward_grad，另外有对应的测试么

Comment (cxxly) [reply python/paddle/incubate/autograd/functional.py]: 是的，我补充下测试用例

File: python/paddle/fluid/tests/unittests/autograd/test_autograd_functional_prim.py (hunk 1 above)
Comment (cxxly) [reply python/paddle/fluid/tests/unittests/autograd/test_autograd_functional_prim.py]: 之前调试windows机器注释的，已经去掉

File: python/paddle/incubate/autograd/functional.py (hunk 2 above)
Comment (cxxly) [reply python/paddle/incubate/autograd/functional.py]: Done

**Review by levi131 — APPROVED** [review]
LGTM

**Review by XieYunshen — APPROVED** [review]
LGTM 
单测重命名

File: python/paddle/incubate/autograd/primapi.py [hunk 4]
```diff
@@ -14,28 +14,26 @@
...
 @framework.static_only
-def forward_gradients(targets, inputs, input_gradients=None):
+def forward_grad(outputs, inputs, grad_inputs=None):
     """Forward mode of automatic differentiation.
 
     .. note::
         **ONLY available in the static mode and primitive operators.**
 
     Args:
-        targets: The target tensor or tensors
```
Comment (TCChenlong) [inline python/paddle/incubate/autograd/primapi.py:31]: 参考下方，需要给出数据类型

File: python/paddle/incubate/autograd/primapi.py [hunk 5]
```diff
@@ -90,4 +88,95 @@ def forward_gradients(targets, inputs, input_gradients=None):
...
+            to None, in this case is equivalent to all ones .
+
+    Returns:
+        grad_inputs (Tensor|Tensors): The gradients for inputs. 
+
+    Examples:
+
+        .. code-block:: python
+
+            import numpy as np
```
Comment (TCChenlong) [inline python/paddle/incubate/autograd/primapi.py:115]: 示例代码注意换行 加一些注释

File: python/paddle/incubate/autograd/primapi.py [hunk 6]
```diff
@@ -90,4 +88,95 @@ def forward_gradients(targets, inputs, input_gradients=None):
     ad = primx.Transform(ys[0].block)
     _, ys_dot = ad.linearize(xs, ys, xs_dot)
 
-    return ys_dot[0] if isinstance(targets, framework.Variable) else ys_dot
+    return ys_dot[0] if isinstance(outputs, framework.Variable) else ys_dot
+
+
+@framework.static_only
+def grad(outputs, inputs, grad_outputs=None):
+    """Reverse mode of automatic differentiation.
```
Comment (TCChenlong) [inline python/paddle/incubate/autograd/primapi.py:96]: Reverse？Backward？

**Review by TCChenlong — APPROVED** [review]

File: python/paddle/incubate/autograd/primapi.py (hunk 4 above)
Comment (cxxly) [reply python/paddle/incubate/autograd/primapi.py:31]: fix with a new pull request.

File: python/paddle/incubate/autograd/primapi.py (hunk 5 above)
Comment (cxxly) [reply python/paddle/incubate/autograd/primapi.py:115]: fix with a new pull request.

File: python/paddle/incubate/autograd/primapi.py (hunk 6 above)
Comment (cxxly) [reply python/paddle/incubate/autograd/primapi.py:96]: fix with a new pull request.

**Review by XiaoguangHu01 — APPROVED** [review]
LGTM

File: python/paddle/incubate/autograd/primapi.py (hunk 6 above)
Comment (cxxly) [reply python/paddle/incubate/autograd/primapi.py:96]: reverse mode,  refs: https://arxiv.org/pdf/1502.05767.pdf

**Pull request merged**